node/src/components/chainspec_loader.rs

```diff
@@ -298,6 +425,224 @@ impl ChainspecLoader {
+                    previous_block_header.timestamp(),
+                    previous_block_header.next_block_era_id(),
+                    initial_pre_state.next_block_height(),
+                    PublicKey::System,
+                );
+
+                // TODO: figure out whether the upgrade changes the validator set based on the
+                // global state and don't pass the previous header if it does.
+                self.execute_immediate_switch_block(
+                    effect_builder,
+                    Some(previous_block_header),
+                    initial_pre_state,
+                    finalized_block,
+                )
+            }
+            Err(error) => {
+                error!("failed to upgrade contract runtime: {}", error);
+                self.reactor_exit = Some(ReactorExit::ProcessShouldExit(ExitCode::Abort));
+                Effects::new()
+            }
+        }
+    }
+
+    /// Creates a switch block after an upgrade or genesis. This block has the system public key as
+    /// a proposer and doesn't contain any deploys or transfers. It is the only block in its era,
+    /// and no consensus instance is run for era 0 or an upgrade point era.
+    fn execute_immediate_switch_block<REv>(
+        &self,
+        effect_builder: EffectBuilder<REv>,
+        maybe_previous_block_header: Option<Box<BlockHeader>>,
+        initial_pre_state: ExecutionPreState,
+        finalized_block: FinalizedBlock,
+    ) -> Effects<Event>
+    where
+        REv: From<ContractRuntimeRequest>
+            + From<StorageRequest>
+            + From<MarkBlockCompletedRequest>
+            + Send,
+    {
+        let protocol_version = self.chainspec.protocol_version();
+        async move {
+            let block_and_execution_effects = effect_builder
+                .execute_finalized_block(
+                    protocol_version,
+                    initial_pre_state,
+                    finalized_block,
+                    vec![],
+                    vec![],
+                )
+                .await?;
+            // We need to store the block now so that the era supervisor can be properly
+            // initialized in the participating reactor's constructor.
+            effect_builder
+                .put_block_to_storage(block_and_execution_effects.block.clone())
+                .await;
+            effect_builder
+                .mark_block_completed(block_and_execution_effects.block.height())
+                .await;
+            info!(
+                immediate_switch_block = ?block_and_execution_effects.block.clone(),
+                "immediate switch block after upgrade/genesis stored"
+            );
+            Ok(block_and_execution_effects)
+        }
+        .event(move |result| Event::ExecuteImmediateSwitchBlockResult {
+            maybe_previous_block_header,
+            result,
+        })
+    }
+
+    fn handle_execute_immediate_switch_block_result<REv>(
```

Comment:
This makes me think that we're stretching the responsibility of the `chainspec_loader` a bit too much. In the future (definitely not as a part of this PR) we should consider introducing dedicated component to handle upgrades and stuff.